### PR TITLE
GH1618: Enhance the Teamcity Provider's PullRequest detection

### PR DESCRIPTION
**Pull request by nmbro**
Implements a lot of #1618

### CODE REVIEW

**Comment by nmbro** [comment]
Tested working in my Teamcity environment using Github

File: src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs [hunk 1]
```diff
@@ -11,6 +11,39 @@ namespace Cake.Common.Build.TeamCity.Data
     /// </summary>
     public class TeamCityPullRequestInfo : TeamCityInfo
     {
+        private static bool InferIsPullRequest(string gitReferenceName)
+        {
+            var branchSlices = gitReferenceName.Split('/');
```
Comment (devlead) [inline src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs]: Should check & return if `gitReferenceName` is null or empty.

File: src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs [hunk 2]
```diff
@@ -11,6 +11,39 @@ namespace Cake.Common.Build.TeamCity.Data
...
+                    default:
+                        return false;
+                }
+            }
+
+            return false;
+        }
+
+        private static int? GetPullRequestNumber(string gitReferenceName)
+        {
```
Comment (devlead) [inline src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs]: Should check & return if `gitReferenceName` is null or empty.

File: src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs (hunk 1 above)
Comment (nmbro) [reply src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs]: done

File: src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs (hunk 2 above)
Comment (nmbro) [reply src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs]: done

**Comment by nmbro** [comment]
@devlead done! ;)

**Comment by martin-pritchard** [comment]
Hi @nmbro / @devlead. Just ran into this bug, and found this PR. What's outstanding? Could see the requested changes, but they appear to have been addressed.

**Pull request merged**